text/3389-manifest-lint.md

```diff
@@ -0,0 +1,398 @@
+- Feature Name: `manifest-lint`
+- Start Date: 2023-02-14
+- RFC PR: [rust-lang/rfcs#3389](https://github.com/rust-lang/rfcs/pull/3389)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Add a `[lints]` table to `Cargo.toml` to configure reporting levels for
+rustc and other tool lints.
+
+# Motivation
+[motivation]: #motivation
+
+Currently, you can configure lints through
+- `#[<level>(<lint>)]` or `#![<level>(<lint>)]`, like `#[forbid(unsafe)]`
+  - But this doesn't scale up with additional targets (benches, examples,
+    tests) or workspaces
+- On the command line, like `cargo clippy -- --forbid unsafe`
+  - This puts the burden on the caller
+- Through `RUSTFLAGS`, like `RUSTFLAGS=--forbid=unsafe cargo clippy`
+  - This puts the burden on the caller
+- In `.cargo/config.toml`'s `target.*.rustflags`
+  - This couples you to the running in specific directories and not running in
+    the right directory causes rebuilds
+  - The cargo team has previously stated that
+    [they would like to see package-specific config moved to manifests](https://internals.rust-lang.org/t/proposal-move-some-cargo-config-settings-to-cargo-toml/13336/14?u=epage)
+
+We would like a solution that makes it easier to share across targets and
+packages for all tools.
+
+See also
+- [rust-lang/rust-clippy#1313](https://github.com/rust-lang/rust-clippy/issues/1313)
+- [rust-lang/cargo#5034](https://github.com/rust-lang/cargo/issues/5034)
+- [EmbarkStudios/rust-ecosystem#59](https://github.com/EmbarkStudios/rust-ecosystem/issues/59)
+- [Proposal: Cargo Lint configuration](https://internals.rust-lang.org/t/proposal-cargo-lint-configuration/9135)
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+A new `lints` table would be added to configure lints:
+```toml
+[lints.rust]
+unsafe = "forbid"
+```
+and `cargo` would pass these along as flags to `rustc` and `clippy`.
+
+This would work with
+[RFC 2906 `workspace-deduplicate`](https://rust-lang.github.io/rfcs/2906-cargo-workspace-deduplicate.html?highlight=2906#):
+```toml
+[lints]
+workspace = true
+
+[workspace.lints]
+unsafe = "forbid"
+```
+
+## Documentation Updates
+
+## The `lints` section
+
+*as a new ["Manifest Format" entry](https://doc.rust-lang.org/cargo/reference/manifest.html#the-manifest-format)*
+
+Override the default level of lints from different tools by assigning them to a new level in a
+table, for example:
+```toml
+[lints.rust]
+unsafe = "forbid"
+```
+
+This is short-hand for:
+```toml
+[lints.rust]
+unsafe = { level = "forbid", priority = 0 }
+```
+
+`level` corresponds to the lint levels in `rustc`:
+- `forbid`
+- `deny`
+- `warn`
+- `allow`
+
+`priority` is a signed integer that controls which lints or lint groups override other lint groups:
```

Comment:
Yeah I suggested this in https://github.com/rust-lang/rfcs/pull/3389#discussion_r1107386273 as well

Reply:
I think the "in the order they are specified" model is _useful_ for when lint levels are being specified in multiple places. But I think it might be fine to do it just for CLI flags.


Strawman proposal: `rustc` gains a `sort-lint-cli-flags` arg that Cargo passes in, which perhaps becomes the default mode in the next edition.

Reply:
Does the `warnings` meta-group complicate this at all? (AFAICT with some quick testing of `-Dwarnings` behavior it ignores flag order).

Reply:
It's not actually a lint group; it turns _all_ warnings into errors, including non-lint warnings

Reply:
> Strawman proposal: `rustc` gains a `sort-lint-cli-flags` arg that Cargo passes in, which perhaps becomes the default mode in the next edition.

I like that better, than adding completely new, but almost identical flags for this.

Reply:
@GilShoshan94 I would ask that we keep the conversation more focused.  There is another recent thread on `level = [lint]`.  As for not using sub-tables, please open a new thread for that.  This is focused on `priority` and it will make things complicated to cover other topics.

Reply:
@GilShoshan94, it is not about priority of the source where the lint is allowed, but about priority in the `Cargo.toml` file itself. E.g.

```toml
[lints.clippy]
allow = ["pedantic"]
warn = ["bool_to_int_with_if"]
```

should it be passed to `rustc`/`clippy` as `-Aclippy::pedantic -Wclippy::bool_to_int_with_if` or `-Wclippy::bool_to_int_with_if -Aclippy::pedantic`. In the first case `bool_to_int_with_if` would warn, in the second case it would not. In this case it is obvious that the user intend is the former, but cargo doesn't know that `pedantic` is a group and overwrites `bool_to_int_with_if`. This gets even more complicated when multiple (overlapping) lint groups are involved.

The question is, can we come up with a heuristic that feels natural to the user or do we need a `pedantic` configuration?

Reply:
I'd expect a certain level of specificity here.

1. Order by location.  The latest one in the series takes precedence.
  1. Read from ~/.cargo/config.toml (if exists)
  2. Read from ./.cargo/config.toml (if exists)
  3. Read from workspace:Cargo.toml (if exists)
  4. Read from crate:Cargo.toml
  5. Read from inlined attributes
2. Within each location, override by specificity. `-Aclippy::pedantic` is less specific than `-Wclippy::bool_to_int_with_if`.  Therefore, irrespective of order, `-Wclippy::bool_to_int_with_if` should be at WARN, and the rest of pedantic should be at ALLOW.

If cargo doesn't know that pedantic is a group, it shouldn't be that hard to add that as a notation and order as such.

Reply:
@spmadden auto-sorting the lint groups is the proposal being discussed in this thread.  One of the questions at hand with the heuristic approach is how to deal with lint groups that intersect, like `rust-2018-idioms` and `rust-2021-compatibility`.

Reply:
Auto-sorting and heuristics do not have to be mutually exclusive.  We can start with one now and add the other later.  The basic idea is that until auto-sorting is added, the order within a priority is unspecified.  Once auto-sorting is added,we'll auto-sort within a priority level.  For this to work, we'd need
- the schema to support adding `priority` (likely `lint = level`)
- the CLI flag to accept a group of lints w/ levels to sort so we can pass it for each priority level
  - I suspect this would already be needed so that we don't auto-sort between lint level sources, keeping the existing priority system working for those sources

I lean towards starting with priority because its a lower level, lower policy mechanism allowing this to be get in sooner and we can iterate on policy ideas for auto-sorting over time in nightly.

Granted, there might be personal biases in there.
- I'm unsure if its originally my idea but I put it in the RFC (was originally proposed in a cargo team meeting I think)
- Its a simpler,. easy to implement, and isolated change rather than to work across rustc, clippy, and rustdoc to get buy-in on the flag and the heuristics, then implement it across all the tools, and then get the enough testing on the heuristics to be satisfied
- I'm familiar with cargo so I can more quickly get the needed changes in rather than having to dig into unfamiliar code or hope someone else picks it up

Reply:
> we can iterate on policy ideas for auto-sorting over time in nightly.

I think this is a good idea. Let's go with the `priority` attribute first and while this is in nightly we can try to find a auto-sorting policy. If we manage to come up with an auto-sorting policy that gets rid of the `priority` attribute, then that would be the best case.

So I recommend to resolve this, by adding a note about the `priority` attribute that it is subject to removal, before the feature gets stabilized and then reference a Future Possibility about an auto-sorting feature.

Reply:
>  My hope is to stablize this sooner than later; to decouple the auto-sort policy work from getting this into people's hands on stable.

My concern is, that if the `priority` attribute gets stabilized with this feature, we can't get rid of it anymore, even when we figure out a good auto-sorting policy. This might put long-term maintenance burden on the `cargo` team. But I don't know much about the cargo code base or how much hassle that is. And I won't have to maintain it, so I'll leave that decision for you.

Reply:
We talked about this in the cargo team meeting. 

We are concerned locking down future decisions based on current policy and implementations.  Examples
- Could clippy change its mind about groups being disjoint?
- We haven't heard from rustc diagnostic or rustdoc folks yet on this (I'll be reaching out once I'm done writing this up)
- We have two likely new tools on the horizon (`cargo`, `cargo-semver-check`) and we don't want to make decisions for them when they haven't yet explored their problem space

Of the options listed below, I feel we are leaning towards
- Priority now, with the possibility of a mixture of auto-sorting and warnings in the future
- The RUSTFLAGS-like schema of `[ {level=lint} ]`

The plan for moving this forward is
- See where this conversation goes
- Get more information from rustc/rustdoc folks
- Ping some specific individuals who have dealt with some of these problems to make sure they have the opportunity to provide feedback

## Options

### Priority only

🔴 We are concerned with a users ability to get it right on the first try or debug it if they didn't.

### Priority with a future possibility for warning when the order doesn't make sense 

**Variant:** Instead of just warn, auto-sort within a priority level (most-specific last)

Example warnings:
- groups being in the same priority as lints within their grup
- intersecting groups in the same priority)

Why warning and not error:
- Only being a warning means previously disjoint groups can be changed to intersect without being a breaking change
- Only being a warning means we can add it later without being a breaking change

We talked about two ways for detecting warning scenarios
- Add new CLI contract for batching the lints by priority level so the tool can make the relevant decisions
- Cargo queries the linter for the lint-group graph and caches it
  - More complexity on cargo
  - Consistent experience that the cargo team can evolve,. rather than each linter implementing it

We can use the time while only priority is stable to double-check our assumption of the user confusion before committing to the future possibility

This allows us to have a fact path to a stable MVP to use

See above for a lot of the details

✅ While `priority` isn't a great experience, holistically this might be acceptable

### Auto-sort with a heuristic

🔴 We are concerned about the fact that this effectively discourages intersecting groups and would prefer the above option as it has the user break ties rather than trying to guess on behalf of the user on how to break ties.

### Use a schema more akin to RUSTFLAGS

Self-reflection: for any of these variants, are we falling into the trap of thinking something is "right" because we started with the CLI / rust-attribute solution (path-dependence)?

**Variant 1: Array of Tables**
```toml
[lints]
# inline
rust = [
    { lint = "all", level = "warn" },
    { lint = "allow_dead_code", level = "allow" },
]
# standard
[[lints.clippy]]
lint = "all"
level = "warn"
[[lints.clippy]]
lint = "cyclomatic_complexity"
level = "allow"
```
- 🔴 We are not thrilled with the syntax for this

**Variant 2: Allow a DSL shorthand**
```toml
[lints]
rust = [
    # short-flag prefix
    "Wall",
    "Aallow_dead_code",
    # more explicit
    "A=allow_dead_code",
    # Mirror the CLI
    "-Aallow_dead_code",
    "--warning=cyclomatic_complexity",
    # with lint config
    { lint = "cyclomatic_complexity", level = "warn", threshold = 50 }
]
```
- 🔴 The short-letter prefix isn't too clear
  - People are used to `Wall` but `Aallow_dead_code` is harder to parse
  - Even with people used to `-Wall`, its become a term of its own to the point that some people do not realize its short for `-W all` or `--warning all`
  - Generally short-flags are meant for humans, long flags are encouraged automation
- 🔴 Not thrilled with any of these as a DSL

**Variant 3: Mirror RUSTFLAGS in TOML**
```toml
[lints]
rust = [
    {allow="all"},
    {warn="allow_dead_code"},
]

clippy = [
    {warn="pedantic"}
]
```
- ✅ This has the feel of RUSTFLAGS without a DSL
- `level=lint` keeps the keys finite so we can add lint configuration in the future

### User-declared groups

```toml
[lints.clippy.groups]
pedantic = "warn"

[lints.clippy.lints]
unused_self = "allow"
```
- Effectively `groups` and `lints` are priority level keywords with everything in `groups` going before `lints`
- 🔴 `all` would be hard coded to go before all other groups
- 🔴 no way to solve intersecting groups
- 🔴 limits group hierarchies
- 🔴 Unless we also implement the same mechanism used for the Warning option, users could put things under the wrong priority level keyword
  - Is it big enough of a deal for us to do anything about?


*phew* I think thats it?

Reply:
Is deserializing `lints.tool` as `IndexMap<LintName, Level>` an option? As in, it's just an object with name to level mapping, but the order inside the TOML determines what overrides?

That is how many CLIs work: if you pass `--no-foo --foo`, `foo` is on, for `--foo, --no-foo` it's off.

Reply:
> * Could clippy change its mind about groups being disjoint?

Highly unlikely, but not impossible. Also other tools (rustc) already do this or will do this.

---

1. Agreed

2. I'm a bit confused: If we can detect conflicts in priority groups and warn about them, shouldn't it then be also possible to resolve those conflicts? Is resolving this always possible automagically? I think having `priority` in the MVP will give us time to figure those things out

3. Agreed

4. I hate everything about this :sweat_smile: : This is just not useable. It's all the bad parts from the `lint = "level"` and the `level = [lint]` syntaxes combined IMO.

5. is maybe an option, but if we punt this to the user, I think `priority` is the better approach.

Reply:
> Is using order for priority an option¹? As in, it's just an object with name to level mapping, but the order inside the TOML determines what overrides.
> 
> That is how many CLIs work: if you pass `--no-foo --foo`, `foo` is on, for `--foo, --no-foo` it's off.
> 
> ¹ easy to maintain in deserialization w/ `IndexMap`

By making the reference to `IndexMap`, I assume you are meaning the order within a table?  The cargo team ruled that out because (1) order for keys is not defined in the TOML spec and we don't want to rely on an extension of the TOML spec and (2) this doesn't communicate semantics to the user/tool and they could easily do things incorrectly, e.g. sorting the table.

Option 4 tries to instead explore ordering within TOML syntax and semantics.

Reply:
> 2 I'm a bit confused: If we can detect conflicts in priority groups and warn about them, shouldn't it then be also possible to resolve those conflicts? 

We could except for when groups intersect.  Framed differently, its the auto-sort solution but punts to the user instead of using a heuristic, with an MVP that always punts to the user.

Something I forgot to include in discussing is going past the MVP is the cargo/linter contract and acknowledging the parallel work for registering custom lints.  I've not followed that too closely.  Depending on how likely that is to be completed and what options are possibilities, each of these proposals could have a limiting factor on the other.

Reply:
I've updated the RFC to reflect Option 2 (`priority` with warnings/auto-sort in the future).